gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXRegionState.java

```diff
@@ -153,7 +153,14 @@ public TXRegionState(final LocalRegion r, final TXState tx) {
     this.expiryReadLock = r.getTxEntryExpirationReadLock();
     this.isValid = true;
 
-    if (!r.isInitialized() && r.getImageState().lockPendingTXRegionStates(true, false)) {
+
+    // check if lock already taken by the same thread
+    // in that case don't take lock
+    ImageState imgState = r.getImageState();
+
+    boolean alreadyLocked = imgState.isWriteLockedBySameThread();
+    if (!r.isInitialized()
+        && (alreadyLocked || imgState.lockPendingTXRegionStates(true, false))) {
       try {
```

Comment:
Since the lock can be taken multiple times by same thread, won't it be better to make this a ReentrantReadWriteLock instead of "isWriteLockedBySameThread" check? The block inside will add TXRegionState in pending list and initialize pendingTXOps lists which I think is required.